packages/grid/_modules_/grid/hooks/features/pagination/useGridPagination.ts

```diff
@@ -2,98 +2,144 @@ import * as React from 'react';
 import { GridApiRef } from '../../../models/api/gridApiRef';
 import { GridPaginationApi } from '../../../models/api/gridPaginationApi';
+import { GridOptions } from '../../../models/gridOptions';
 import { GridPageChangeParams } from '../../../models/params/gridPageChangeParams';
 import { useGridApiOptionHandler } from '../../root/useGridApiEventHandler';
 import { useGridApiMethod } from '../../root/useGridApiMethod';
 import { optionsSelector } from '../../utils/optionsSelector';
+import { useGridState } from '../core/useGridState';
 import { gridContainerSizesSelector } from '../../root/gridContainerSizesSelector';
 import { useLogger } from '../../utils/useLogger';
-import { useGridReducer } from '../core/useGridReducer';
 import { useGridSelector } from '../core/useGridSelector';
 import { visibleGridRowCountSelector } from '../filter/gridFilterSelector';
-import {
-  GRID_INITIAL_PAGINATION_STATE,
-  GridPaginationActions,
-  gridPaginationReducer,
-  GridPaginationState,
-  setGridPageActionCreator,
-  setGridPageSizeActionCreator,
-  setGridPaginationModeActionCreator,
-  setGridRowCountActionCreator,
-} from './gridPaginationReducer';
+import { GridPaginationState } from './gridPaginationState';
 
-const PAGINATION_STATE_ID = 'pagination';
+function applyConstraints(
+  state: GridPaginationState,
+  options: { page: GridOptions['page'] },
+): GridPaginationState {
+  const pageCount =
+    state.pageSize && state.rowCount > 0 ? Math.ceil(state.rowCount / state.pageSize!) : 1;
+
+  return {
+    ...state,
+    pageCount,
+    page: Math.min(pageCount - 1, options.page !== undefined ? options.page : state.page),
+  };
+}
 
 export const useGridPagination = (apiRef: GridApiRef): void => {
   const logger = useLogger('useGridPagination');
 
-  const { dispatch } = useGridReducer<GridPaginationState, GridPaginationActions>(
-    apiRef,
-    PAGINATION_STATE_ID,
-    gridPaginationReducer,
-    { ...GRID_INITIAL_PAGINATION_STATE },
-  );
+  const [, setGridState, forceUpdate] = useGridState(apiRef);
   const options = useGridSelector(apiRef, optionsSelector);
   const visibleRowCount = useGridSelector(apiRef, visibleGridRowCountSelector);
   const containerSizes = useGridSelector(apiRef, gridContainerSizesSelector);
 
   const setPage = React.useCallback(
     (page: number) => {
       logger.debug(`Setting page to ${page}`);
-      dispatch(setGridPageActionCreator(page));
 
-      // we use getState here to avoid adding a dependency on gridState as a dispatch change the state, it would change this method and create an infinite loop
-      const params: GridPageChangeParams = apiRef.current.getState<GridPaginationState>(
-        PAGINATION_STATE_ID,
+      setGridState((state) => ({
+        ...state,
+        pagination: applyConstraints(
+          {
+            ...state.pagination,
+            page,
+          },
+          {
+            page: options.page,
```

Comment:
you can just pass options here

Reply:
x2

Reply:
option.page is a dependency in the effect/callback. Using option as a dependency would mean it rerun a lot more often.

Reply:
ah yes that would create issues.
or just pass the page directly as arg as you don't have other options and it's a scoped to this file...

Reply:
Yeah, we could provide the page directly. I have used options to retain the same semantic, to make it obvious that it's coming from the props, but happy to refactor.

Reply:
Ok I ve done the change for you. So we get it in our latest release ;)